### DISCUSSION TITLE
Netflix Conductor TaskResult and Worker client

### DISCUSSION THREAD

**Discussion by manishkushwaha1412** [Netflix/conductor#2959] [General]
Hi Friends, 

I am trying to write netflix conductor client and following document given in netflix conductor how-tos (below url).
https://netflix.github.io/conductor/how-tos/Workers/build-a-java-task-worker/

I am using gradle and could see its not recognising TaskClient and while search in google I don't get much help as well. I can execute workflow from conductor UI (from workbench). However, not able to execute it though API (API executes successful but no result displayed in UI, one issue also created for the same - https://github.com/Netflix/conductor/issues/2942). I thought I would call it thought java client but when I started writing Worker dependency for TaskClient is not resolved. Can someone suggest or help me here please. 

I appreciate response here. 

Thanks
Manish

**Comment by jxu-nflx** [comment]
@manishkushwaha1412 Can you please elaborate your issue here? As it's not clear to me of whether you are having TaskClient issue (as a result, the workflow tasks are not being picked up and processed)? Or are you facing search issue on the UI (as you describe in the other thread https://github.com/Netflix/conductor/issues/2942)?

**Reply by jxu-nflx** [reply]
For the TaskClient issue, if you can post your depeency file and some code snippet, that would help to identity issues. For the UI search issue, it would be helpful if you could post the responses from both ui and swagger for the same query.

**Reply by manishkushwaha1412** [reply]
@jxu-nflx thanks for your response, please find details below

Issue 1: I am trying to call workflow though API, API executes successfully but I could not see any result in the UI, this is a search issue and ticket is already created for the same (mentioned in above). 

Issue 2: 
There are couple of ways I can call workflow. 1. thought API, 2 thought netflix conductor UI workbench and thought Worker also. API is not working so I am trying with netflix conductor ui workbench http://localhost:5000/workbench. This UI is working fine if workflow is triggered from mentioned URL. Now I am trying with Worker option where though java program I will call worker. In order to do that I need to write WorkerClient as mentioned in below URL 
https://netflix.github.io/conductor/how-tos/Workers/build-a-java-task-worker/
However, when I am writing Worker client, it uses TaskResult which is not recognised in dependency. Please look at screen grab below. 
![image](https://user-images.githubusercontent.com/62668735/166627426-3ec1b2e4-cf7b-414f-b331-8e7a88c738d8.png)

whereas I could see library is already added to the classpath but still it is not recognised. I thought it could be IDE issue but I tried this in IntelliJ and Eclipse both and getting the same issue 
![image](https://user-images.githubusercontent.com/62668735/166627559-0864418c-e57a-4485-9f0f-cf5b1455bf9e.png)

Hope this is clear now. Please suggest.

**Reply by jxu-nflx** [reply]
Let's focus on your issue # 2 first.

> API is not working so I am trying with netflix conductor ui workbench http://localhost:5000/workbench. This UI is working fine if workflow is triggered from mentioned URL.
What you describe here is confusing to me, as what the UI does is simply call the conductor API to start the workflow. So it doesn't make sense that the conductor API is not working but UI works fine.

Now, about your Worker client, are you using maven or gradle. Do you have any dependency lock files in your project? If so, you might want to update your dependency lock after you added the `conductor-client` library. Also try build your project from command line instead of intellij, as something intellij could be mis-configured.

**Reply by manishkushwaha1412** [reply]
**In the first question**: there are two ways I can trigger a workflow 1. through URL  http://localhost:5000/workbench and 2. I can call API which is running in server i.e. localhost:8080/api/workflow - this is post request and I need to send workflow name and input to the workflow. My Question here is - when I am trying through URL  http://localhost:5000/workbench I could see workflow is invoked and I can see result also in http://localhost:5000/execution/{workflowId}. However, I am trying to call API in 8080 its executing but I don't see result any where. Means, go through https://netflix.github.io/conductor/labs/running-first-workflow/ and try to create tasks, workflow and call workflow and check if you can see details anywhere in the UI. I think its clear now else I can connect with you and explain issues what I am facing. 

**In the second question**: I tried in both maven and gradle but issue is same. I don't have any dependency lock file not sure why I am not getting TaskResult. I am building my project from command line, below dependency I have added in my gradle project 
```
dependencies {
    implementation group: 'com.netflix.conductor', name: 'conductor-client', version: '3.3.4'
    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.8.1'
    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.8.1'
}
```
sample code snap below 
![image](https://user-images.githubusercontent.com/62668735/168567606-9185ace7-4bb9-4722-b519-3b3078b97678.png)

**Reply by jxu-nflx** [reply]
Hello @manishkushwaha1412 so the `call workflow` issue is now resolved? Would be happy to connect to walk you through if you are still facing issues.
For the dependency issue, what is the error message when you build your worker project via console? i.e. ./gradlew clean build? Also what is the java version you are using?

**Reply by manishkushwaha1412** [reply]
Please let me know your availability so that we can connect, I am not able to build conductor locally even through docker-build as well.

**Comment by manishkushwaha1412** [comment]
@jxu-nflx 
I have one more question. I wanted to connect conductor to my local database mysql/mongo. I could see corresponding connection configurations are available like `mysql-persistence` `postgres-persistence` etc. I am not getting file where I will keep my local database properties. could you please help me with that.

**Reply by jxu-nflx** [reply]
For postgres/mysql issues, please post it here: https://github.com/Netflix/conductor-community So you can get better assistant as we don't use postgres/mysql for conductor internally

**Reply by v1r3n** [reply]
@manishkushwaha1412 the connection details for the database will be in the same file that contains conductor properties.  Conductor server loads this file using this system property:
https://github.com/Netflix/conductor/blob/main/server/src/main/java/com/netflix/conductor/Conductor.java#L50

**Comment by csotiriou** [comment]
I am having this exact issue regarding the inability to use Task and TaskResult, nothing from `conductor-commons` is getting recognized when importing `conductor-client` , and this seems to be a packaging issue.

I have attached a very simple reproducer.

[conductor-dependency-issue.zip](https://github.com/Netflix/conductor/files/9868846/conductor-dependency-issue.zip)


The interesting thing is that this issue appeared in version 3.0.0 onwards. If I use version 2.31.9 and below the `Task` and `TaskResult` classes are getting recognized properly.

The attached project is in Quarkus, but I can reproduce the same behaviour with Spring and any stack whatsoever. I would appreciate any input on how to solve this. @manishkushwaha1412 what was the fix about the dependency issue in your case?

**Reply by csotiriou** [reply]
EDIT: I found the issue. It seems we need to also get the conductor-commons explicitly and add it in the POM or Gradle file. Even though it's a transient dependency, it will not be automatically included in the project. This is in stark contrast to what was happening with 2.x, or other libraries in general. 

By adding an explicit dependency to the `conductor-commons` library, the problem was fixed. Maybe this should be mentioned in the documentation?
